monitor/server.cpp

```diff
@@ -20,146 +20,219 @@
+		}());
 
-	std::string reply = make_reply(req, content);
-	async_write(reply);
+		m_collect_ts = std::chrono::system_clock::now();
+
+		async_write();
+
+	});
 }
 
-void handler::async_write(std::string data) {
+void handler::async_write() {
+	DNET_LOG_DEBUG(m_monitor.node(), "monitor: http-server: send requested statistics: started: {}, size: {}",
+	               m_remote, m_response.size());
+
 	auto self(shared_from_this());
-	m_report = std::move(data);
-	DNET_LOG_DEBUG(m_monitor.node(), "monitor: server: send requested statistics: started: {}:{}, size: {}",
-	               m_remote, m_socket.remote_endpoint().port(), m_report.size());
-	boost::asio::async_write(m_socket, boost::asio::buffer(m_report),
-	                         std::bind(&handler::handle_write, self));
-}
 
-void handler::handle_write() {
-	DNET_LOG_DEBUG(m_monitor.node(), "monitor: server: send requested statistics: finished: {}:{}", m_remote,
-	               m_socket.remote_endpoint().port());
-	close();
+	boost::asio::async_write(m_socket, boost::asio::buffer(m_response), [self,
+	                                                                     this](const boost::system::error_code &err,
+	                                                                           const long unsigned int &size) {
+		const auto finish_ts = std::chrono::system_clock::now();
+
+		if (err) {
+			DNET_LOG_ERROR(m_monitor.node(), "monitor: http-server: failed to send response: {}",
+			               err.message());
+			return;
+		}
+
+		DNET_LOG_DEBUG(m_monitor.node(), "monitor: http-server: send requested statistics: finished: {}",
+		               m_remote);
+
+		const auto full_url = [this]() {
+			const auto &endpoint = m_socket.local_endpoint();
+			const bool ipv6 = endpoint.protocol() == boost::asio::ip::tcp::v6();
+
+			std::ostringstream stream;
+			stream << "http://" << (ipv6 ? "[" : "") << endpoint.address().to_string() << (ipv6 ? "]" : "")
+			       << ":" << endpoint.port() << m_url;
+			return stream.str();
+		}();
+
+		DNET_LOG_INFO(m_monitor.node(), "monitor: http-server: client: {}, url: \"{}\", response-size: {}, "
+		                                "recv-time: {} usecs, collect-time: {} usecs, send-time: {} usecs, "
+		                                "total-time: {} usecs",
+		              m_remote, full_url, size,
+		              std::chrono::duration_cast<std::chrono::microseconds>(m_recv_ts - m_start_ts).count(),
+		              std::chrono::duration_cast<std::chrono::microseconds>(m_collect_ts - m_recv_ts).count(),
+		              std::chrono::duration_cast<std::chrono::microseconds>(finish_ts - m_collect_ts).count(),
+		              std::chrono::duration_cast<std::chrono::microseconds>(finish_ts - m_start_ts).count());
+
+		close();
+	});
 }
 
 void handler::close() {
-	boost::system::error_code ec;
-	m_socket.shutdown(boost::asio::socket_base::shutdown_both, ec);
+	m_socket.shutdown(boost::asio::socket_base::shutdown_both);
```

Comment:
This change may lead to throwing an exception. Previous version was exception-free, as desired.